### PR TITLE
Add github action file, to update tables of contents automatically

### PR DESCRIPTION
**Pull request by claremacrae**
See the original issue #4367, and #4372, which this was a part of.

## Changes made:

1. Added a GitHub Actions file that will trigger the updating of tables of contents markers - and if they are changed, will create a new commit and push

## Things to document, somewhere, for users...

* The tool
  * The tool used to generate the tables of contents ("toc") is [MarkdownSnippets](https://github.com/simonCropp/MarkdownSnippets) - or mdsnippets
  * The tools options are configured in `mdsnippets.json`
  * conan-center-index is only using its toc capability, and not its embedding of example code
* Writing docs
    * To add a table of contents in a new file, place single line `toc` near the start of the .md file - before a level-2 heading (`##`)
    * Headings must be formatted with `##` and not underlines

## Notes on the GitHub Action file

* It runs as user `actions-user`, which will need to ["sign the CLA"](https://github.com/conan-io/conan-center-index/pull/4371#issuecomment-767826930)!
* It only runs on changes committed to master
* It won't create a new commit if nothing changed
* The commit message is in the file - and is `[docs] Regenerate tables of contents`
* Once it has run, the output will be viewable at https://github.com/conan-io/conan-center-index/actions
* It might be worth adding a build badge for that. I'm happy to do that, if wanted...

### CODE REVIEW

**Comment by claremacrae** [comment]
I'm closing this, 3 weeks on... I've been able to update the tables of contents by running mdsnippets manually, using the runes described in #4594 - and I think that there is too much discussion and work involved in add this automation mechanism....

**Comment by Croydon** [comment]
Some decisions take a long time since there is always a lot to do. I don't think you should have closed this already before getting feedback from the CCI team @jgsogo @danimtb @SSE4 @uilianries

**Comment by claremacrae** [comment]
> Some decisions take a long time since there is always a lot to do. I don't think you should have closed this already before getting feedback from the CCI team ...

OK thanks, well if it's decided that this is worth doing, I would change it to supply the configuration on the command line, to remove the separate file from cluttering up the CCI root directory.

**Comment by jgsogo** [comment]
Hi, @claremacrae 

Sorry for the delay, but sometimes it is not up to us what we can use or not... and we need to ask (and get an answer back). **Now we know we can use Github Actions for these _small_ automations**, and we will love to see them. Having them in actions instead of external systems (Jenkins, JFrog Pipelines) will make it easier to maintain and contribute to them.... and overall, make it possible for you or anyone else to keep improving the automation (there are some bots around this repo that can be added as Github actions as well).

So, if you feel like it, you can open the PR again and we will review and eventually merge (CCI bot will complain, don't worry about it).

**Comment by claremacrae** [comment]
Hi @jgsogo - thanks very much indeed!!

I've just updated the PR to remove the need for a separate configuration file.

To test the new code, I added it to another repo - and it works... Here is the commit made by the bot.

https://github.com/claremacrae/commandline-videostore-cpp/commit/3675645d92b22e8dcf280c00597cd888efcea75e

Remaining steps I'm aware of:

* [ ] Get the bot to avoid the "sign the Conan CLA"...
* [ ] Check that the yml file has the correct branch on its trigger
* [ ] Document somehow that when new markdown pages are added, they should have a `toc` line added after the first heading, which is the placeholder of the table of contents

File: .github/workflows/on-push-do-doco.yml
Comment (jgsogo) [inline .github/workflows/on-push-do-doco.yml:17]:
```suggestion
          --exclude-directories 'recipes' \
```

Not sure if it works this way, but we need to exclude the folder with the recipes, just in case there is a `.md` file there. If modified it will modify the recipe revision and we would need to build it again.

File: .github/workflows/on-push-do-doco.yml
Comment (jgsogo) [inline .github/workflows/on-push-do-doco.yml:26]: I'm having a look and the best I've found so far have been this SO question (https://stackoverflow.com/questions/58033366/how-to-get-current-branch-within-github-actions). I can't believe there is no easier way to get the branch name 😅 

Maybe `${GITHUB_REF#refs/heads/}` is more clear.

**Review by jgsogo — CHANGES_REQUESTED** [review]
Good! Need to validate the `recipes` directory exclusion. 

@danimtb , can you run the command in Windows and check that the `--exclude-directories 'recipes'` will prevent modifying files in the `recipes` folder?

File: .github/workflows/on-push-do-doco.yml
Comment (claremacrae) [reply .github/workflows/on-push-do-doco.yml:26]: I've just run it on Mac, and it does... (the mdsnippets tool is available on Mac, Linux and Windows - the CI job uses windows as that's what the author users...


```mdsnippets --convention InPlaceOverwrite --exclude-directories 'recipes' --toc-level 5
Config:
    TargetDirectory: /Users/clare/Documents/develop/Conan/conan-center-index-claremacrae
    UrlPrefix: 
    LinkFormat: GitHub
    Convention: InPlaceOverwrite
    TocLevel: 5
    ValidateContent: False
    HashSnippetAnchors: False
    TreatMissingAsWarning: False
    FileConfigPath:  (exists:False)
    ExcludeDirectories:
        recipes
Found 4 files for snippets
MarkdownProcessor Finished. 49ms
Finished 283ms
```

I know from past runs that there's no way that this would have taken only 283 ms if it had traversed recipes...

**Comment by claremacrae** [comment]
> Good! Need to validate the `recipes` directory exclusion.
> 
> @danimtb , can you run the command in Windows and check that the `--exclude-directories 'recipes'` will prevent modifying files in the `recipes` folder?

I've just run it on Mac, and it does... (the mdsnippets tool is available on Mac, Linux and Windows - the CI job uses windows as that's what the author users...


```mdsnippets --convention InPlaceOverwrite --exclude-directories 'recipes' --toc-level 5
Config:
    TargetDirectory: /Users/clare/Documents/develop/Conan/conan-center-index-claremacrae
    UrlPrefix: 
    LinkFormat: GitHub
    Convention: InPlaceOverwrite
    TocLevel: 5
    ValidateContent: False
    HashSnippetAnchors: False
    TreatMissingAsWarning: False
    FileConfigPath:  (exists:False)
    ExcludeDirectories:
        recipes
Found 4 files for snippets
MarkdownProcessor Finished. 49ms
Finished 283ms
```

I know from past runs that there's no way that this would have taken only 283 ms if it had traversed recipes...

File: .github/workflows/on-push-do-doco.yml
Comment (claremacrae) [reply .github/workflows/on-push-do-doco.yml:26]: I'm unsure if you need any input from me at this stage on the GITHUB_REF comments... 

I just took the YAML code straight from the author of the mdsnippets tool: 

https://github.com/SimonCropp/MarkdownSnippets/blob/master/docs/on-push-do-docs.yml

Comment (claremacrae) [reply .github/workflows/on-push-do-doco.yml:26]: On second thoughts, I did some experimenting - using info from https://stackoverflow.com/a/59780579/104370

Which pointed to this:
https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context

In this run.... https://github.com/claremacrae/commandline-videostore-cpp/runs/1922585010?check_suite_focus=true

I had this step:

```yml
      - name: Check Branch references
        run: |
          echo "github.ref = ${{ github.ref }}"
          echo "github.head_ref = ${{ github.head_ref }}"
          echo "github.base_ref = ${{ github.base_ref }}"
          echo "GITHUB_REF = $GITHUB_REF"
          echo "GITHUB_HEAD_REF = $GITHUB_HEAD_REF"
          echo "GITHUB_BASE_REF = $GITHUB_BASE_REF"
```

And the output was this - where `starting-point` was the name of the branch:

```
Run echo "github.ref = refs/heads/starting-point"
  echo "github.ref = refs/heads/starting-point"
  echo "github.head_ref = "
  echo "github.base_ref = "
  echo "GITHUB_REF = $GITHUB_REF"
  echo "GITHUB_HEAD_REF = $GITHUB_HEAD_REF"
  echo "GITHUB_BASE_REF = $GITHUB_BASE_REF"
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
github.ref = refs/heads/starting-point
github.head_ref = 
github.base_ref = 
GITHUB_REF = 
GITHUB_HEAD_REF = 
GITHUB_BASE_REF = 
```

So `${{ github.ref }}` looks like the right answer to me...

This is based on the belief that you don't have to worry about references relating to pull requests, as the script deliberately only acts on push to master:

```yml
on:
  # Only update the docs on default branch, to reduce the chance of conflicts
  push:
    branches:
      - master
```

**Comment by danimtb** [comment]
I am taking a look at this. My intention is to give it a try on our staging env to verify that actions work well with our jenkins pipeline and the flow in the master branch is correct. Thank you!

**Comment by danimtb** [comment]
Hi @claremacrae. I have been testing the action in our staging environment and unfortunately we would need to change the strategy. At first sight, pushing the `mdsnippets` to master branch seemed the best approach but as the branch is protected the action cannot push changes there.

As an alternative, I have modified the action to push the `mdsnippets` changes to the PR branch before it is merged. I think this would require the "Allow edits from maintainers" check in the PR and could lead to conflicts if more than one PR updates the docs. This is the action I used:

```
name: docs_markdown_toc
on:
  pull_request:
    paths:
      - 'docs/**'
jobs:
  docs_markdown_toc:
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v2
    - name: Run MarkdownSnippets
      run: |
        dotnet tool install --global MarkdownSnippets.Tool
        mdsnippets ${GITHUB_WORKSPACE} \
          --convention InPlaceOverwrite \
          --exclude-directories 'recipes' \
          --toc-level 5
      shell: bash
    - name: Push changes
      run: |
        git config --local user.email "action@github.com"
        git config --local user.name "GitHub Action"
        git commit -m "[docs] Regenerate tables of contents" -a || echo "nothing to commit"
        remote="https://${GITHUB_ACTOR}:${{secrets.GITHUB_TOKEN}}@github.com/${GITHUB_REPOSITORY}.git"
        echo ${GITHUB_HEAD_REF}
        branch="${GITHUB_HEAD_REF}"
        git push "${remote}" ${branch} || echo "nothing to push"
      shell: bash
```

Another approach would be to make the action create a PR from a predefined branch in the repo whenever the TOCs need to be updated. This could also lead to conflicts or race conditions but would be less "intrusive" for contributors (does not modify their PRs).

Let me know what you think and if you could give the action above a go from your side. Thanks!

**Comment by Croydon** [comment]
> Another approach would be to make the action create a PR from a predefined branch in the repo whenever the TOCs need to be updated. This could also lead to conflicts or race conditions but would be less "intrusive" for contributors (does not modify their PRs).

I think this is a much better solution. Also, if only the already merged documentation counts and it pushes always to the same branch/PR, it can't create conflicts

**Comment by danimtb** [comment]
After @Croydon comment, I reconsidered the "pull request from repo branch" approach and got it working with the following action:

*docs_markdown_toc.yml*
```
name: docs_markdown_toc
on:
  push:
    branches:
      - master
jobs:
  docs_markdown_toc:
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v2
    - name: Run MarkdownSnippets
      run: |
        dotnet tool install --global MarkdownSnippets.Tool
        mdsnippets ${GITHUB_WORKSPACE} \
          --convention InPlaceOverwrite \
          --exclude-directories 'recipes' \
          --toc-level 5
      shell: bash
    - name: Create Pull Request
      uses: peter-evans/create-pull-request@v3
      with:
        branch: action-doc-toc
        commit-message: "[docs] Regenerate tables of contents"
        title: "[docs] Regenerate tables of contents"
        body: "Automatic update of the documentation TOCs."
```

**Comment by claremacrae** [comment]
> After @Croydon comment, I reconsidered the "pull request from repo branch" approach and got it working with the following action:

It looks great to me...

What happens next? How do we moved forward with this? I'd be happy for you to create a new PR with that improved content....
Or should I copy your change in to this PR?

**Comment by danimtb** [comment]
@claremacrae I think it is fine to make the changes in this PR. We made changes in the staging pipeline to allow contributions to the `.github` folder (for github actions, bot configuration...) so that the bot doesn't complain. It will be ready for production this week and we could finally merge this PR with the TOC automation 🥳

**Comment by claremacrae** [comment]
> @claremacrae I think it is fine to make the changes in this PR. We made changes in the staging pipeline to allow contributions to the `.github` folder (for github actions, bot configuration...) so that the bot doesn't complain. It will be ready for production this week and we could finally merge this PR with the TOC automation 🥳

Thanks... I just added that code to the PR - with @danimtb as a co-author ...

**Comment by danimtb** [comment]
Perfect @claremacrae! I will take care of merging this PR when we deploy the new changes to production (hopefully this week) :smile:

**Comment by danimtb** [comment]
With the latest changes in the pipeline, CI is now 🟢 and PRs like this one are tagged with `GitHub config` label. Now it can be reviewed and it will be merged automatically by the bot 🌈

**Review by danimtb — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Comment by claremacrae** [comment]
> With the latest changes in the pipeline, CI is now 🟢 and PRs like this one are tagged with `GitHub config` label. Now it can be reviewed and it will be merged automatically by the bot 🌈

Huge thanks @danimtb - and everyone else who has given feedback on this... 🎉  !

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**
